File: Problem-Solving/LeetCode/0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL) return NULL;
        vector<ListNode*> vt;
        ListNode* node = head;
        while (1) {
            vt.push_back(node);
            if (node->next == NULL) break;
            node = node->next;   
        }
        ListNode* newHead = vt[vt.size()-1];
        for(int i=vt.size()-1;i>=1;i--) {
            vt[i]->next = vt[i-1];
        }
        vt[0]->next = NULL;
        return newHead;
    }
};

File: Problem-Solving/LeetCode/0206-reverse-linked-list/0206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    // [1, 2] -> [2, 1]
    // [] -> []
    if head == nil {
        return nil
    }
    nodes := make([]*ListNode, 0)
    node := head
    for {
        nodes = append(nodes, node)
        if node.Next == nil {
            break;
        }
        node = node.Next
    }    
    newHead := nodes[len(nodes)-1]
    for i := len(nodes)-1;i>=1;i-- {
        node = nodes[i]
        node.Next = nodes[i-1]
    }
    nodes[0].Next = nil
    return newHead
}